src/dataframe/dataframe.jl

```diff
@@ -472,249 +463,67 @@ function upgrade_scalar(df::DataFrame, v::Any)
-    setindex!(df, v, row_inds, findall(col_inds))
-end
-function Base.setindex!(df::DataFrame,
-                        v::AbstractVector,
-                        row_inds::AbstractVector{<:Real},
-                        col_inds::AbstractVector{<:ColumnIndex})
-    for col_ind in col_inds
-        insert_multiple_entries!(df, v, row_inds, col_ind)
+    for (j, col) in enumerate(idxs)
+        df[row_inds, col] = new_df[!, j]
     end
     return df
 end
 
-# df[MultiRowIndex, MultiColumnIndex] = Single Item
-function Base.setindex!(df::DataFrame,
-                        v::Any,
-                        row_inds::AbstractVector{Bool},
-                        col_inds::AbstractVector{Bool})
-    setindex!(df, v, findall(row_inds), findall(col_inds))
-end
+# df[MultiRowIndex, MultiColumnIndex] = AbstractMatrix
 function Base.setindex!(df::DataFrame,
-                        v::Any,
-                        row_inds::AbstractVector{Bool},
-                        col_inds::AbstractVector{<:ColumnIndex})
-    setindex!(df, v, findall(row_inds), col_inds)
-end
-function Base.setindex!(df::DataFrame,
-                        v::Any,
-                        row_inds::AbstractVector{<:Real},
-                        col_inds::AbstractVector{Bool})
-    setindex!(df, v, row_inds, findall(col_inds))
-end
-function Base.setindex!(df::DataFrame,
-                        v::Any,
-                        row_inds::AbstractVector{<:Real},
-                        col_inds::AbstractVector{<:ColumnIndex})
-    for col_ind in col_inds
-        insert_multiple_entries!(df, v, row_inds, col_ind)
+                        mx::AbstractMatrix,
+                        row_inds::Union{AbstractVector, Not},
+                        col_inds::Union{AbstractVector, Regex, Not})
+    idxs = index(df)[col_inds]
+    if size(mx, 2) != length(idxs)
+        throw(DimensionMismatch("number of selected columns ($(length(idxs))) and a" *
+                                " matrix ($(size(mx, 2))) do not match"))
+    end
+    for (j, col) in enumerate(idxs)
+        df[row_inds, col] = view(mx, :, j)
     end
     return df
 end
 
-# df[rows, Regex] = value
-Base.setindex!(df::DataFrame, v::Any, row_inds, col_inds::Regex) =
-    setindex!(df, v, row_inds, index(df)[col_inds])
-
-# df[:] = DataFrame, df[:, :] = DataFrame
-function Base.setindex!(df::DataFrame,
-                        new_df::DataFrame,
-                        row_inds::Colon,
-                        col_inds::Colon=Colon())
-    setfield!(df, :columns, copy(_columns(new_df)))
-    setfield!(df, :colindex, copy(index(new_df)))
-    df
-end
-
-# df[:, :] = ...
-Base.setindex!(df::DataFrame, v, ::Colon, ::Colon) =
-    (df[1:size(df, 1), 1:size(df, 2)] = v; df)
-
-# df[Any, :] = ...
-Base.setindex!(df::DataFrame, v, row_inds, ::Colon) =
-    (df[row_inds, 1:size(df, 2)] = v; df)
-
-# df[:, Any] = ...
-Base.setindex!(df::DataFrame, v, ::Colon, col_inds) =
-    (df[col_inds] = v; df)
-
```

Comment:
The ability to remove so many setindex! methods here is awesome!

Reply:
There will be some new methods after the deprecation period, but for sure much less.

The problem with the old rules was that `:` was treated specially in different situations. For example `df[:, :] = df2` removed all columns from `df` and inserted into it columns from `df2`.

Now we have a clear rule `:` for row selection is not special. It is the same as `1:nrow(df)` so things get simpler (and are consistent with Base). For special operations we have `!` as row selector (so when one uses it this means that one knows that it is non-standard).